### DISCUSSION TITLE
Usage without React (Vanilla JS client)

### DISCUSSION THREAD

**Discussion by markusc** [wevm/wagmi#361] [Q&A] [answered]
I was really excited to try wagmi in a SvelteKit app since this [tweet](https://twitter.com/awkweb) and the [documentation](https://wagmi.sh/docs/client) stated that it could now be used without React. However, it seems to still require React to function.

The following in a SvelteKit app:
```
import { createWagmiClient } from 'wagmi'
    
const client = createWagmiClient()
```

Results in:

![image](https://user-images.githubusercontent.com/23310708/165248665-47125551-da29-4f0e-9f25-3d6c0bab3d1c.png)

Did I misunderstand or is it possible to use the client in a non-React app?

**Comment by tmm** [comment, marked as answer]
@markusc really excited for you to try this out! For non-React, you need to use `@wagmi/core` instead of `wagmi`.

```svelte
import { createClient } from '@wagmi/core'
    
const client = createWagmiClient()
```

There's not a lot in the way of documentation for the Vanilla JS route yet. I recommend taking a look at the [`core` package source](https://github.com/tmm/wagmi/tree/main/packages/core/src) and test cases (e.g. [`src/actions/connect.test.ts`](https://github.com/tmm/wagmi/blob/main/packages/core/src/actions/accounts/connect.test.ts)). They are mostly self-documenting and coverage a lot of the surface area. There is also an old [`using-core`](https://github.com/tmm/wagmi/tree/fca6b1b8a3644ec6f08d50073d2ceda39fbd4c3a/examples/using-core) example that uses the Vanilla client in React.

Feel free to create a new discussion topic for your adventures with SvelteKit!

**Reply by ennioVisco** [reply]
> @markusc really excited for you to try this out! For non-React, you need to use `@wagmi/core` instead of `wagmi`.
> 
> ```svelte
> import { createClient } from '@wagmi/core'
>     
> const client = createWagmiClient()
> ```
> 
> There's not a lot in the way of documentation for the Vanilla JS route yet. I recommend taking a look at the [`core` package source](https://github.com/tmm/wagmi/tree/main/packages/core/src) and test cases (e.g. [`src/actions/connect.test.ts`](https://github.com/tmm/wagmi/blob/main/packages/core/src/actions/accounts/connect.test.ts)). They are mostly self-documenting and coverage a lot of the surface area. There is also an old [`using-core`](https://github.com/tmm/wagmi/tree/fca6b1b8a3644ec6f08d50073d2ceda39fbd4c3a/examples/using-core) example that uses the Vanilla client in React.
> 
> Feel free to create a new discussion topic for your adventures with SvelteKit!

Hello, thanks for the nice work. This library is a pleasure to learn to use and explore.

I'm trying to include it in a Nuxt 3/Vue 3 project as a Vanilla JS library. The setup seem to work fine, however I noticed that I am missing 
some basic hooks in `@wagmi/core` (e.g. `useAccount`). Is there any specific reason for that? Or should they just be imported from react?

**Reply by jxom** [reply]
You can use `getAccount` & `watchAccount` from `@wagmi/core` :)

**Reply by ennioVisco** [reply]
Update on how I'm using it in Nuxt 3, since I have been contacted by a fellow developer struggling to integrate it.

Essentially it can be summarized in a few steps:

### 1. Package installation
```sh
npm install @wagmi/core@0.10.11
```
Please note that my setup is not compatible with 1.x wagmi, which is super cool, but I still didn't have the time to migrate to the new apis!

### 2. Plugin setup
The following lines are required to make wagmi available at runtime:

```ts
// plugins/wagmi.ts
import {
  configureChains,
  createClient,
  goerli,
  InjectedConnector,
  mainnet,
} from "@wagmi/core";
// import { polygon, polygonMumbai } from "@wagmi/chains";  // other chains are available here...
import { infuraProvider } from "@wagmi/core/providers/infura";

export default defineNuxtPlugin(() => {
  const apiKey = useRuntimeConfig().public.infuraId; // fetches the infura key from the .env NUXT_PUBLIC_INFURA_ID variable
  const supportedChains = [mainnet, goerli]; // other chains must be added here
  const { provider, webSocketProvider } = configureChains(supportedChains, [
    infuraProvider({ apiKey }),
  ]);

  const client = createClient({
    autoConnect: true,
    provider,
    webSocketProvider,
    connectors: [new InjectedConnector()],
  });
  return { provide: { wagmi: () => client } };
});
```

### 3. Composables
We found it useful to define the following composables:

```ts
// composables/blockchain.ts
import {
  connect,
  getAccount,
  InjectedConnector,
  prepareWriteContract,
  watchAccount,
  GetAccountResult,
  writeContract,
  switchNetwork,
  mainnet,
  goerli,
} from "@wagmi/core";
// import { polygon, polygonMumbai } from "@wagmi/chains"; // other chains available here
import { Abi } from "@fyblo/http-types/SmartContract";

export type Chain = "Ethereum" | "Goerli" ; //  But maybe also... | "Matic" | "Mumbai";

export type ContractInfo = {
  address: `0x${string}`;
  abi: Abi;
  chain: Chain;
};

export const useChainAccount = () => {
  const chainAccount = useState<GetAccountResult>("chainAccount", () =>
    getAccount(),
  );

  if (!chainAccount.value.isConnected) {
    watchAccount(() => {
      chainAccount.value = getAccount();
    });
  }

  return chainAccount;
};

export const useConnect = async () => {
  const currentAccount = getAccount();
  if (!currentAccount.isConnected) {
    const connection = await connect({
      chainId: goerli.id,
      connector: new InjectedConnector(),
    });
    console.log("Connecting to:", connection);
  } else {
    throw new Error("Already connected");
  }
};

export const switchChain = async (chain: Chain) => {
  switch (chain) {
    case "Ethereum":
      return await switchNetwork({
        chainId: mainnet.id,
      });
    case "Goerli":
      return await switchNetwork({
        chainId: goerli.id,
      });
     // Similar cases for other chains...
  }
};

export const useContract = async <T>(contractInfo: ContractInfo) => {
  async function execute(functionName: string, args: unknown[]) {
    if (!getAccount().isConnected) {
      await useConnect();
    }
    switchChain(contractInfo.chain);

    const request = await prepareWriteContract({
      address: contractInfo.address,
      abi: contractInfo.abi,
      functionName,
      args,
    });

    console.log("Request:", request);

    return await writeContract(request);
  }

  return {
    execute,
  };
};
```

### 4. Practical usage
Here is an example of how we use it:

```vue
<script lang="ts" setup>
import type { Chain } from "~/composables/blockchain";

// Contract info...
const contractAddress = "0x...";
const contractAbi = "..." as const;
const contractChain = "Ethereum";

// Vue refs...
const wallet = ref("");
const tokenAmount = ref(0);

const doIssue = async () => {
  try {
    const { execute } = await useContract({
      address: contractAddress,
      abi: contractAbi,
      chain: contractChain as Chain,
    });
    return execute("mint", [wallet.value, tokenAmount.value]);
  } catch (error) {
    console.log("Token issuing error: ", error);
  }
};
</script>

<template>
    <form @submit="submitClubDealInvite">
         <input type="text" v-model="wallet" />
         <input type="number" v-model="tokenAmount" />
         <button type="submit">Send</button>
    <form>
<template>
```

Hope it helps!

**Comment by matthewhausman** [comment]


**Reply by matthewhausman** [reply]
not currently sure how to replicate the query persistence, you guys r using the 'persistQueryClient' function from @tanstack/react-query-persist-client... not sure what the vue equivalent would be. For another day !

**Reply by ennioVisco** [reply]
> not currently sure how to replicate the query persistence, you guys r using the 'persistQueryClient' function from @tanstack/react-query-persist-client... not sure what the vue equivalent would be. For another day !

What about [`useStorage`](https://vueuse.org/core/usestorage/) or [`useCached`](https://vueuse.org/core/usecached/#usecached)?

**Reply by matthewhausman** [reply]
https://github.com/matthewhausman/vue-ethereum-hooks

after looking into this there's actually a core package that's in vanilla js which is working great.

theres already a pretty complete pkg that's called vagmi, but this one is going to be a simpler and more similar to the react hooks due to some recent contributions i made to tanstack and some simple adjustments to wagmi core

**Reply by ennioVisco** [reply]
So why not include the Vue hooks directly in wagmi?
The new code shouldn't be a lot, right?

**Reply by matthewhausman** [reply]
I would think it's super doable relative to the type of tasks that the wagmi team is accustomed to working on. 

Not sure if its a great time in the scope of the roadmap however. Going to talk to the team again when the unit testing / playground functionality is implemented in the repo above in an identical way to their repo, will hopefully help for us to get first class support

**Reply by reslear** [reply]
I think it’s worth thinking about it, since there is already a successful experience for example from https://github.com/DamianOsipiuk/vue-query --> https://github.com/TanStack/query/tree/main/packages/vue-query

**Reply by matthewhausman** [reply]
btw.. been using a mangled version of wagmi hooks for vue for a startup I'm doing but will migrate over the hooks im using today. the repo above only has a couple hooks because I had been stuck on issues related to pnpm / publishing

**Reply by reslear** [reply]
What are the problems? I can help

**Reply by tmm** [reply]
Excited to see this conversation! Definitely interested in adapters for many frameworks (Vue probably the first).

There's a little bit more to execute on the roadmap that will make this easier and we don't want to increase surface area for the library too much before then, but this is definitely a priority for the core team in 2023. Until then, using `@wagmi/core` or a community package might be the easiest.

**Reply by matthewhausman** [reply]
> What are the problems? I can help

actually managed to solve them. Latest pkg version has a few hook examples. 

u can import the wagmiVue object from the pkg and init with

`wagmiVue.install(app, {
    autoConnect: true,
    provider,
    webSocketProvider,
    connectors: [
      new MetaMaskConnector({ chains }),
      new CoinbaseWalletConnector({
        chains,
        options: {
          appName: 'moonlight.xyz',
        },
      }),
      new WalletConnectConnector({
        chains,
        options: {
          qrcode: true,
        },
      }),
      new InjectedConnector({
        chains,
        options: {
          name: 'Injected',
          shimDisconnect: true,
        },
      }),
    ],
  })`

 the last big thing is to make some adjustments to tanstack's vue query 'useQuery' composable so that we can match the hook APIs even more similarly.  After that it's smooth sailing for the hook implementation. I could use help for sure though msg me on twitter @iceman_1738

**Reply by reslear** [reply]
@tmm it would be cool, at least to start, to publish a package with 1-2 basic methods implemented, and the community will support the rest

**Reply by reslear** [reply]
@matthewhausman write issue with problems in you repo

**Reply by matthewhausman** [reply]
@reslear good idea - will notify here when it's available

**Reply by matthewhausman** [reply]
https://github.com/matthewhausman/vue-ethereum-hooks/discussions/3
